src/metrics/metrics.ts

```diff
@@ -350,3 +357,138 @@ export function getMateAndAssists(pgnMoveLine: string) {
     lastPieceMoved,
   };
 }
+
+export async function getGameWithMostMoves(games: FileReaderGame[]) {
+  let maxNumMoves = 0;
+  let gameWithMostMoves: FileReaderGame | null = null;
+  let gameLinkWithMostMoves = null;
+
+  for await (const game of games) {
+    const chess = new Chess();
+
+    chess.loadPgn(game.moves);
+    const numMoves = chess.history().length;
+
+    if (numMoves > maxNumMoves) {
+      maxNumMoves = numMoves;
+      gameWithMostMoves = game;
+      let site = game.metadata
+      .find((item) => item.startsWith('[Site "'))
+      ?.replace('[Site "', '')
+      .replace('"]', '');
+      gameLinkWithMostMoves = site;
+    }
+  }
+
+  return {
+    gameLinkWithMostMoves,
+    maxNumMoves,
+  };
+
+}
+
+export async function getPieceLevelMoveInfo(games: FileReaderGame[]) {
+  const numMovesByPiece = {};
+  let averageNumMovesByPiece = {};
+  let pieceWithMostMovesInAGame = null;
+  let pieceWithHighestAverageNumMoves = null;
+  let gameLinkWithPieceMostMoves = null;
+  let numMovesMadePieceWithMostMoves = 0;
+  
+  let gameCount = 0;
+
+  for (const game of games) {
+    gameCount++;
+    const chess = new Chess();
+    const moveGenerator = chess.historyGenerator(game.moves);
+
+    const numMovesByPieceThisGame = {};
+
+    // update move counts of each unambiguous piece
+    for (let moveInfo of moveGenerator) {
+      const { move } = moveInfo;
+
+      let movedPiece = move.unambiguousSymbol;
+
+      if (movedPiece) {
+        if (!numMovesByPiece[movedPiece]){
+          numMovesByPiece[movedPiece] = 0
+        }
+        numMovesByPiece[movedPiece]++;
+        numMovesByPieceThisGame[movedPiece]++
+
+        // Check if the move is a castling move
+        if (moveInfo.move.flags === 'k' || moveInfo.move.flags === 'q') {
+          let movingRook;
+    
+          if (moveInfo.move.flags === 'k') {
+            movingRook = 'rh';
+          } else {
+            movingRook = 'ra';
+          }
+    
+          if (moveInfo.move.color === 'w') {
+            movingRook = movingRook.toUpperCase();
+          }
+
+          if (!numMovesByPiece[movingRook]){
+            numMovesByPiece[movingRook] = 0
+          }
+
+          // duplicative action for the array capturing moves by piece for this game
+          if (!numMovesByPieceThisGame[movingRook]){
+            numMovesByPieceThisGame[movingRook] = 0
+          }
```

Comment:
if you want to save all these checks you can use a `Map` or safe dictionary. Will make the code more readable

Reply:
good 2 know. can do this later

Reply:
this is minor, just for readability